lib/services/database_service.dart

```diff
@@ -12,37 +12,30 @@ import 'package:adventures_in_chat_app/extensions/extensions.dart';
 class DatabaseService {
   String currentUserId;
   final Database _database;
-  final _controllers = <String, StreamController<List<MessagesListItem>>>{};
 
   DatabaseService({Database database})
       : _database = database ?? FirestoreDatabase();
 
   Stream<List<MessagesListItem>> getMessagesStream(String conversationId) {
-    // fresh controller per unique conversation.
-    StreamController<List<MessagesListItem>> _controller;
-    if (_controllers.containsKey(conversationId)) {
-      _controller = _controllers[conversationId];
-    } else {
-      _controller = StreamController<List<MessagesListItem>>();
-    }
-
-    _database.getMessagesStream(conversationId).listen((messagesList) {
-      var latest = DateTime.fromMicrosecondsSinceEpoch(0); // init to minimum
-      final newList = <MessagesListItem>[];
+    return _database
+        .getMessagesStream(conversationId)
+        .transform<List<MessagesListItem>>(
+            StreamTransformer.fromHandlers(handleData: (messagesList, sink) {
+      var latest =
+          messagesList.isNotEmpty ? messagesList.first.timestamp : null;
```

Comment:
Starting with the first doesn't work well for the first day with multiple messages unless we initialize the itemsList to latest like we had it before. Though we wouldn't want to init the list with the first item null. 

<img width="930" alt="Screen Shot 2020-07-06 at 6 26 34 am" src="https://user-images.githubusercontent.com/829044/86541620-a9306a00-bf51-11ea-8420-8d7d2699d61f.png">

Initialising the latest to the first possible DateTime is a solid algorithm and it is better the way it was.

Reply:
I don't understand what you're saying - what is better the way what was when?  Are you saying your previous solution is better than this one?  I have a reason I did it this way and would be happy to discuss and happy to come to the best solution we can agree on together but I'm not going to respond to "it was better the way it was" with an incredibly unclear description of why, instead of asking me why I did it the way I did it that way and inviting discussion. I'm pretty over this tbh, I'm taking this down and getting on with other work.

Reply:
Sorry @nickmeinhold reading that again I understand that I was a little sleep deprived this morning and very unclear. What I should have said is it doesn't show the date for the first day. The screen grabs side by side do show what I mean though with the before starting with the date, and the after not.

Initially on Saturday we initialised `newList` to have the first item of `latest` which was better though I wasn't sure if we would get away with initialising the list with a first item of null. That will make your solution work.

```dart
final newList = <MessagesListItem>[latest];
```

Reply:
Thanks, I think I understand the problem now.  You're using `newList` but I'm assuming you mean `itemsList`?  

`latest` is a timestamp so I'm guessing you mean: 
```
final newList = <MessagesListItem>[SectionDate(latest)];
```
? 

As you say, probably worth avoiding the null, what do you think about: 
```
final itemsList = messagesList.isNotEmpty ? <MessagesListItem>[SectionDate(latest)] : <MessagesListItem>[];
```

Avoiding the null is getting a bit messy, might be better to check at the top of the function and handle null in one go, what do you think?  I'll try it and test manually and see how it goes and report back.

Reply:
Perfect that will fix it. Nice work. Sorry of the confusion earlier.

Reply:
Thanks mate, sorry my response was a bit stroppy, I appreciate your response and that we're able to quickly get back on track.  Reporting back: I tested the latest change manually and none of the previous issues were present and the date was there for the first day... go team!